### PR TITLE
ROX-10731: wire postgres SAC search for alerts

### PR DESCRIPTION
**Pull request by rhybrillou**
## Description

Ensure scoped access control (data filtering) is performed correctly for alert search when running against postgres.

## Checklist
- [ ] Investigated and inspected CI test results
~~- [ ] Unit test and regression tests added~~
~~- [ ] Evaluated and added CHANGELOG entry if required~~
~~- [ ] Determined and documented upgrade steps~~
~~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

## Testing Performed

Integration tests are being added with PR https://github.com/stackrox/stackrox/pull/871 (in the scope of ticket ROX-9620)
These will validate the behaviour.

### CODE REVIEW

File: proto/storage/alert.proto [hunk 1]
```diff
@@ -17,11 +17,11 @@ message Alert {
...
         string                  type                = 4;
-        string                  namespace           = 5 [(gogoproto.moretags) = 'search:"Namespace"'];
-        string                  namespace_id        = 16 [(gogoproto.moretags) = 'search:"Namespace ID,hidden"'];
+        string                  namespace           = 5;  // This field has to be duplicated in Alert for scope management and search.
+        string                  namespace_id        = 16; // This field has to be duplicated in Alert for scope management and search.
         map<string, string>     labels              = 7;
-        string                  cluster_id          = 9  [(gogoproto.moretags) = 'search:"Cluster ID,hidden"'];
-        string                  cluster_name        = 10 [(gogoproto.moretags) = 'search:"Cluster,hidden"'];
+        string                  cluster_id          = 9;  // This field has to be duplicated in Alert for scope management and search.
+        string                  cluster_name        = 10; // This field has to be duplicated in Alert for scope management and search.
```
Comment (janisz) [inline proto/storage/alert.proto:24]: If you remove search tag how it could be used for search?

Comment (rhybrillou) [reply proto/storage/alert.proto:24]: Alerts exist in three flavours:
- Deployment alerts
- Image alerts
- Resource alerts
For deployment and resource, it's easy to find in which kubernetes cluster and namespace the alert fired. For image alerts, it's a bit more complicated, and currently not really checked.
The search tags being removed here are the ones for deployment alerts.
`Cluster ID` and `Namespace` search fields are now duplicated at the root of the `Alert` proto message when upserting. These fields are located lines 108 to 111 in the current file. The flags are being removed here to reduce the chances of collision on search fields.

File: central/alert/datastore/internal/search/searcher_impl.go [hunk 2]
```diff
@@ -27,7 +28,8 @@ var (
 		Reversed: true,
 	}
 
-	alertSearchHelper = sac.ForResource(resources.Alert).MustCreateSearchHelper(mappings.OptionsMap)
+	alertSearchHelper           = sac.ForResource(resources.Alert).MustCreateSearchHelper(mappings.OptionsMap)
+	alertPosgresSACSearchHelper = sac.ForResource(resources.Alert).MustCreatePgSearchHelper(mappings.PostgresOptionsMap)
```
Comment (md2119) [inline central/alert/datastore/internal/search/searcher_impl.go:32]: Pull the options map from schema

File: central/alert/datastore/internal/search/searcher_impl.go [hunk 3]
```diff
@@ -130,7 +132,13 @@ func convertAlert(alert *storage.ListAlert, result search.Result) *v1.SearchResu
 ///////////////////////////////////////////////
 
 func formatSearcher(unsafeSearcher blevesearch.UnsafeSearcher) search.Searcher {
```
Comment (md2119) [inline central/alert/datastore/internal/search/searcher_impl.go:134]: fork this out. PaginatedSearcher is entirely redundant in postgres, and defaultsortedsearcher should not be required. It should come from clients (service layer or other consumers)

File: proto/storage/alert.proto [hunk 4]
```diff
@@ -17,11 +17,11 @@ message Alert {
         string                  id                  = 1 [(gogoproto.moretags) = "search:\"Deployment ID,store,hidden\" sql:\"index=hash\""];
         string                  name                = 2 [(gogoproto.moretags) = "search:\"Deployment,store\""];
         string                  type                = 4;
-        string                  namespace           = 5 [(gogoproto.moretags) = 'search:"Namespace"'];
```
Comment (md2119) [inline proto/storage/alert.proto:20]: We have same search tag on multiple search fields within alert object. Does this function correctly with bleve? @connorgorman @viswajithiii

File: proto/storage/alert.proto (hunk 1 above)
Comment (connorgorman) [reply proto/storage/alert.proto:24]: Yeah this seems fine

**Review by connorgorman — APPROVED** [review]

**Pull request merged**